File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if head is None or head.next is None:
            return True
        
        stack = []
        fast = head 
        slow = head
        
        while fast and fast.next:
            stack.append(slow.val)
            fast = fast.next.next
            slow = slow.next
            
        if fast: slow = slow.next
            
        while slow:
            if stack.pop() != slow.val:
                return False
            slow = slow.next
        
        return True
        
        

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        std::list<int> ls = {};
        
        while (head!=nullptr){
            int val = head->val;
            ls.push_back(val);
            head = head->next;
        }
        
        std::list<int> rls(ls.rbegin(), ls.rend());
            
        return ls==rls;      
        
    }
};